jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/CreateLambdaFunction.kt

```diff
@@ -6,26 +6,80 @@ package software.aws.toolkits.jetbrains.services.lambda.upload
 import com.intellij.openapi.actionSystem.AnAction
 import com.intellij.openapi.actionSystem.AnActionEvent
 import com.intellij.openapi.actionSystem.LangDataKeys
+import com.intellij.openapi.project.Project
 import icons.AwsIcons
 import software.amazon.awssdk.services.lambda.model.Runtime
+import software.aws.toolkits.jetbrains.services.cloudformation.CloudFormationTemplateIndex
 import software.aws.toolkits.jetbrains.services.iam.IamRole
 import software.aws.toolkits.jetbrains.services.lambda.runtime
 import software.aws.toolkits.jetbrains.services.lambda.upload.EditFunctionMode.NEW
 import software.aws.toolkits.resources.message
 
-class CreateLambdaFunction(private val handlerName: String? = null) : AnAction(message("lambda.create_new"), null, AwsIcons.Actions.LAMBDA_FUNCTION_NEW) {
+open class CreateLambdaFunction() : AnAction(message("lambda.create_new"), null, AwsIcons.Actions.LAMBDA_FUNCTION_NEW) {
     override fun actionPerformed(event: AnActionEvent) {
         val project = event.getRequiredData(LangDataKeys.PROJECT)
         val runtime = event.runtime()
 
-        val dialog = if (handlerName != null) {
-            EditFunctionDialog(project = project, mode = NEW, runtime = runtime, handlerName = handlerName)
-        } else {
-            EditFunctionDialog(project = project, mode = NEW, runtime = runtime)
-        }
+        val dialog = getEditFunctionDialog(project, runtime)
 
         dialog.show()
     }
+
+    protected open fun getEditFunctionDialog(project: Project, runtime: Runtime?): EditFunctionDialog =
+        EditFunctionDialog(project = project, mode = NEW, runtime = runtime)
+}
+
+open class CreateElementBasedLambdaFunction(private val handlerName: String, private val project: Project)
+    : CreateLambdaFunction() {
+    protected var handlerSearch: String = handlerName
+    protected var isExactMatch: Boolean = false
+
+    override fun update(e: AnActionEvent?) {
+        super.update(e)
+
+        val allowAction = CloudFormationTemplateIndex.listFunctions(project)
+            .mapNotNull { it.handler() }
+            .none {
+                if (isExactMatch) {
+                    it == handlerSearch
+                } else {
+                    it.startsWith(handlerSearch)
+                }
+            }
+
+        e?.presentation?.isEnabled = allowAction
+        e?.presentation?.isVisible = allowAction
+    }
+
+    override fun getEditFunctionDialog(project: Project, runtime: Runtime?): EditFunctionDialog =
+        EditFunctionDialog(project = project, mode = NEW, runtime = runtime, handlerName = handlerName)
+}
+
+class CreateLambdaFunctionFromJavaClass(private val handlerName: String, private val project: Project)
+    : CreateElementBasedLambdaFunction(handlerName, project) {
+
+    init {
+        handlerSearch = "$handlerName::"
+        isExactMatch = false
+    }
+}
+
+class CreateLambdaFunctionFromJavaMethod(private val handlerName: String, private val project: Project)
```

Comment:
We shouldn't have any language specific implementations in the core - the only place they should be brought is via explicit extension points.

Reply:
I should have renamed the classes to something more general.

What if I expose the difference in functionality (exact vs startswith) with a class member, and collapse the classes back down?

I do like the idea of separating the most basic `CreateLambdaFunction` used by `plugin.xml` from the rest, to avoid working with nullable types.